### PR TITLE
Add inner tail-recursive definitions

### PR DESCRIPTION
**Pull request by diesalbla**

### CODE REVIEW

File: src/reflect/scala/reflect/internal/Types.scala [hunk 1]
```diff
@@ -3059,12 +3059,15 @@ trait Types
...
-    else underlying match {
-      case ExistentialType(qs, restpe) => newExistentialType(quantified ::: qs, restpe)
-      case _                           => ExistentialType(quantified, underlying)
-    }
+  def newExistentialType(quantified: List[Symbol], underlying: Type): Type = {
+    @tailrec
+    def go(qsyms: List[Symbol], ul: Type): Type =
+      ul match {
+        case ExistentialType(qs, restpe) => go(qsyms ::: qs, restpe)
+        case _                           => ExistentialType(qsyms, underlying)
```
Comment (retronym) [inline src/reflect/scala/reflect/internal/Types.scala]: Should be `ExistentialType(qsyms, ul)`

**Comment by retronym** [comment]
How about just marking the existing methods `@tailrec final`? They aren't designed to be customized in subclasses.

**Review by retronym — CHANGES_REQUESTED** [review]
- Fix typo in code
 - consider just marking the existing methods `@tailrec` final
 - The (rough) convention in the compiler is to call these inner tail rec methods `loop`.

**Comment by retronym** [comment]
I'm taking a broader sweep looking for methods that are already tail recursive and ought to be marked as such and methods (like these) that could be tail recursive with a final modifier:

https://github.com/retronym/scala/tree/topic/tailrec-patrol

**Comment by diesalbla** [comment]
@retronym Regarding marking the methods as `@tailrec` and `final`... can the compiler optimise tail calls in public methods?

**Comment by retronym** [comment]
Yep

**Comment by diesalbla** [comment]
After looking at the changes being done by @retronym  https://github.com/retronym/scala/tree/topic/tailrec-patrol, it seems that branch also adds the improvements originally submitted here. 

That being the case, I am closing this PR in favour of Jason's branch.

**Comment by retronym** [comment]
I've submitted my patch as #7876

**Pull request closed without merging**